operator/controllers/seldondeployment_controller.go

```diff
@@ -1518,13 +1519,20 @@ func (r *SeldonDeploymentReconciler) createDeployments(components *components, i
 						instance.Status.Replicas = found.Status.Replicas
 					}
 				}
-				log.Info("Deployment status ", "name", found.Name, "status", found.Status)
+				log.Info("Deployment status", "name", found.Name, "status", found.Status)
 				if found.Status.ReadyReplicas == 0 || found.Status.UnavailableReplicas > 0 {
 					if ready {
-						condition := getDeploymentCondition(found, appsv1.DeploymentAvailable)
-						log.Info("Updating condition for deployment", "name", found.Name, "condition", condition)
-						instance.Status.SetCondition(machinelearningv1.DeploymentsReady, condition)
+
+						availableCondition := getDeploymentCondition(found, appsv1.DeploymentAvailable)
+						log.Info("Updating availableCondition for deployment", "name", found.Name, "availableCondition", availableCondition)
+						instance.Status.SetCondition(machinelearningv1.DeploymentsReady, availableCondition)
 						log.Info("Inference status", "status", instance.Status)
+
+						progressingCondition := getDeploymentCondition(found, appsv1.DeploymentProgressing)
+						if progressingCondition.IsFalse() && availableCondition.IsFalse() {
+							log.Info("Deployment is not progressing, returning failed status", "name", found.Name)
+							return false, fmt.Errorf("deployment %s is not progressing and not available: (%s)", found.Name, progressingCondition.GetMessage())
```

Comment:
I don't think we want an error here. An error is failed Reconcile which is serious and will cause k8s to resubmit the reconcilliation. Here we simply want to update the instance.Status via a setCondition.

Reply:
Hmm ok, currently the only thing that sets the status to `machinelearningv1.StatusStateFailed` is `updateStatusForError` which is only called with an error so I assumed that was the expected pattern. This seemed to produce the desired result but didn't consider that its a failed Reconcile.

![Screenshot from 2022-01-13 15-05-51](https://user-images.githubusercontent.com/11834492/149509203-9b1f42c4-b8f2-4afd-af84-bbe75ae40117.png)

However if we just set the status as failed here, the main Reconcile loop will just set it as `machinelearningv1.StatusStateCreating` afterwards as it does that for anything where the `deploymentsReady` is false. I guess we'll have to change that too.

So I propose: 
* Not return an error, instead said status to failed and the status description to the error message I've set here
* In the main reconcile loop don't set status to Creating if it's already Failed

Does that make sense?

Reply:
Maybe extend the return type of `createDeployments` to return something more informative like `ready`,`progressing` - so maybe two bools? and then set status based on that?

Reply:
The key is we don't want to re-reconcile. We have done our job and updated the status.